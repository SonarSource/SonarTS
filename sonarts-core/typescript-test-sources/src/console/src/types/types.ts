import {Step} from './gettingStarted'
import {TypedValue} from './utils'

export interface RelayConnection<T> {
  edges: RelayEdge<T>[]
}

interface RelayEdge<T> {
  node: T
}

export interface Viewer {
  id: string
  user: Customer
  model: Model
  project: Project
  crm: any
}

export interface Customer {
  id: string
  name: string
  email: string
  projects: RelayConnection<Project>
  crm: CrmSystemBridge
}

export type SeatStatus = 'JOINED' | 'INVITED_PROJECT' | 'INVITED_CUSTOMER'

export interface Seat {
  name: string | null
  email: string
  isOwner: boolean
  status: SeatStatus
}

export interface CrmSystemBridge {
  id: string
  information: CrmCustomerInformation
  onboardingStatus: CrmOnboardingStatus
  customer: Customer
}

export interface CrmCustomerInformation {
  id: string
  name: string
  email: string
  isBeta: boolean
}

export type Environment = 'Node' | 'Browser'
export type GraphQLClient = 'fetch' | 'lokka' | 'relay' | 'apollo'

export type Example = 'ReactRelay' | 'ReactNativeApollo' | 'ReactApollo' | 'AngularApollo' | 'VueApollo'

export interface CrmOnboardingStatus {
  id: string
  gettingStarted: Step
  gettingStartedSkipped: boolean
  gettingStartedExample: Example | null
}

export interface Project {
  id: string
  name: string
  alias: string
  version: number
  models: RelayConnection<Model>
  relations: RelayConnection<Relation>
  actions: RelayConnection<Action>
  permanentAuthTokens: RelayConnection<PermanentAuthToken>
  authProviders: RelayConnection<AuthProvider>
  integrations: RelayConnection<Integration>
  actionSchema: string
  schema: string
  typeSchema: string
  enumSchema: string
  seats: RelayConnection<Seat>
  projectBillingInformation: ProjectBillingInformation
  region?: string
  packageDefinitions: RelayConnection<PackageDefinition>
  functions: RelayConnection<ServerlessFunction>
}

export interface ServerlessFunction {
  id: string
  name: string
  type?: FunctionType
  binding?: FunctionBinding
  webhookUrl: string
  _webhookUrl?: string
  webhookHeaders: string
  _webhookHeaders?: {[key: string]: string}
  inlineCode?: string
  auth0Id?: string
  logs: RelayConnection<Log>
  stats?: FunctionStats
  modelId?: string
  model?: Model
  operation?: RequestPipelineMutationOperation
  isActive: boolean
  query?: string
}

export interface FunctionStats {
  requestHistogram: number[]
  requestCount: number
  errorCount: number
  lastRequest: Date
}

export type RequestPipelineMutationOperation = 'CREATE' | 'UPDATE' | 'DELETE'

export interface Log {
  id: string
  requestId: string
  duration: number
  status: LogStatus
  timestamp: Date
  message: string
}

export type LogStatus = 'SUCCESS' | 'FAILURE'

export interface RequestPipelineMutationFunction extends ServerlessFunction {
  model: Model
  binding: FunctionBinding
}

export type FunctionType = 'WEBHOOK' | 'AUTH0'

export type FunctionBinding = 'TRANSFORM_ARGUMENT' | 'PRE_WRITE' | 'TRANSFORM_PAYLOAD'

export interface PackageDefinition {
  id: string
  definition: string
  email: string
  name: string
}

export interface ProjectBillingInformation {
  plan: string
}

export interface Integration {
  id: string
  isEnabled: boolean
  name: IntegrationNameType
  type: IntegrationTypeType
}

export interface SearchProviderAlgolia extends Integration {
  applicationId: string
  apiKey: string
  algoliaSyncQueries: RelayConnection<AlgoliaSyncQuery>
  algoliaSchema: string
}

export interface AlgoliaSyncQuery {
  id: string
  indexName: string
  fragment: string
  isEnabled: boolean
  model: Model
}

export type IntegrationNameType =
    'AUTH_PROVIDER_AUTH0'
  | 'AUTH_PROVIDER_DIGITS'
  | 'AUTH_PROVIDER_EMAIL'
  | 'SEARCH_PROVIDER_ALGOLIA'

export type IntegrationTypeType = 'AUTH_PROVIDER' | 'SEARCH_PROVIDER'

export type FieldType =
    'Relation'
  | 'Int'
  | 'String'
  | 'Boolean'
  | 'Enum'
  | 'Float'
  | 'DateTime'
  | 'Password'
  | 'Json'
  | 'GraphQLID'

export interface Field {
  id: string
  name: string
  description: string
  isRequired: boolean
  isList: boolean
  isUnique: boolean
  enumId?: string
  enum?: Enum
  isSystem?: boolean
  isReadonly?: boolean
  typeIdentifier?: FieldType
  defaultValue?: TypedValue
  enumValues: string[]
  reverseRelationField?: Field
  relatedModel?: Model
  relation?: Relation
  model?: Model
  migrationValue?: TypedValue
  constraints?: Constraint[]
  interface?: Interface
}

interface Interface {
  isSystem: boolean
  name: string
}

export interface Relation {
  id: string
  name: string
  isRequired: boolean
  description?: string
  leftModel: Model
  rightModel: Model
  fieldOnLeftModel: Field
  fieldOnRightModel: Field
  permissionSchema: string
  permissions: RelayConnection<RelationPermission>
  permissionQueryArguments: PermissionQueryArgument[]
}

export type UserType = 'EVERYONE' | 'AUTHENTICATED'

export type PermissionRuleType = 'NONE' | 'GRAPH' | 'WEBHOOK'

export interface Enum {
  id: string
  name: string
  values: string[]
}

export interface Model {
  id: string
  name: string
  namePlural: string
  fields: RelayConnection<Field>
  unconnectedReverseRelationFieldsFrom?: Field[]
  itemCount: number
  description: string
  isSystem: boolean
  permissions: RelayConnection<ModelPermission>
  permissionSchema: string
  permissionQueryArguments: PermissionQueryArgument[]
}

export interface PermissionQueryArgument {
  name: string
  typeName: string
  group: string
}

export interface ModelPermission {
  id: string
  fieldIds?: string[]
  ruleWebhookUrl?: string
  rule: Rule
  ruleName?: string
  ruleGraphQuery?: string
  applyToWholeModel: boolean
  isActive: boolean
  operation: Operation
  userType: UserType
}

export interface RelationPermission {
  id: string
  connect: boolean
  disconnect: boolean
  ruleWebhookUrl?: string
  rule: Rule
  ruleName?: string
  ruleGraphQuery?: string
  isActive: boolean
  userType: UserType
}

export type Rule = 'NONE' | 'GRAPH' | 'WEBHOOK'
export type Operation = 'READ' | 'CREATE' | 'UPDATE' | 'DELETE'

export type ActionTriggerType = 'MUTATION_MODEL' | 'MUTATION_RELATION'
export type ActionHandlerType = 'WEBHOOK'

export interface Action {
  id: string
  isActive: boolean
  description: string
  triggerType: ActionTriggerType
  handlerType: ActionHandlerType
  triggerMutationModel?: ActionTriggerMutationModel
  triggerMutationRelation?: ActionTriggerMutationRelation
  handlerWebhook?: ActionHandlerWebhook
}

export type ActionTriggerMutationModelMutationType = 'CREATE' | 'UPDATE' | 'DELETE'

export interface ActionTriggerMutationModel {
  id: string
  fragment: string
  model: Model
  mutationType: ActionTriggerMutationModelMutationType
}

export type ActionTriggerMutationRelationMutationType = 'ADD' | 'REMOVE'

export interface ActionTriggerMutationRelation {
  id: string
  fragment: string
  relation: Relation
  mutationType: ActionTriggerMutationRelationMutationType
}

export interface ActionHandlerWebhook {
  id: string
  url: string
}

export interface User {
  id: string
  createdAt?: string
}

export interface Node {
  id: string
  [key: string]: any
}

export interface PermanentAuthToken {
  id: string
  name: string
  token: string
}

export interface AuthProvider {
  id: string
  type: AuthProviderType
  isEnabled: boolean
  digits: AuthProviderDigits | null
  auth0: AuthProviderAuth0 | null
}

export type AuthProviderType = 'AUTH_PROVIDER_EMAIL' | 'AUTH_PROVIDER_DIGITS'
  | 'AUTH_PROVIDER_AUTH0' | 'anonymous-auth-provider'

export interface AuthProviderAuth0 {
  domain: string
  clientId: string
  clientSecret: string
}

export interface AuthProviderDigits {
  consumerKey: string
  consumerSecret: string
}

export interface OrderBy {
  fieldName: string
  order: 'ASC' | 'DESC'
}

export type FieldWidths = { [key: string]: number }

declare global {
  interface Element {
    scrollIntoViewIfNeeded(centerIfNeeded: boolean): void
  }
}

export interface TetherStep {
  step: Step
  title: string
  description?: string
  buttonText?: string
  copyText?: string
}

export type RelationPopupDisplayState = 'DEFINE_RELATION' | 'SET_MUTATIONS'
export type Cardinality = 'ONE_TO_ONE' | 'ONE_TO_MANY' | 'MANY_TO_ONE' | 'MANY_TO_MANY'

export type BreakingChangeIndicatorStyle = 'TOP' | 'RIGHT'

export type ConstraintType = 'REGEX' | 'CONTAINS' | 'STARTS_WITH' | 'ENDS_WITH' | 'EQUALS' | 'LENGTH'
export type Operator = 'EQ' | 'GT' | 'LT' | 'GTE' | 'LTE'

export interface Constraint {
  type: ConstraintType
  value: string
  lengthOperator?: Operator
}

export type PricingPlan = '2017-02-free' | '2017-02-project' | '2017-02-startup' | '2017-02-growth' | 'enterprise'

export interface PricingPlanInfo {
  name: string
  price: number
  maxStorage: number
  maxRequests: number
  maxSeats: number
  pricePerAdditionalMB: number
  pricePerThousandAdditionalRequests: number
}

export type CreditCardInputDisplayState = 'CREDIT_CARD_DATA' | 'ADDRESS_DATA'

export type Region = 'EU_WEST_1' | 'AP_NORTHEAST_1' | 'US_WEST_2'

export interface PermissionVariable {
  name: string
  typeIdentifier: FieldType
  category?: string
  isRequired: boolean
  isList: boolean
}

export interface Invoice {
  overageRequests: number
  overageStorage: number
  timestamp: string
  total: number
  usageRequests: number[]
  usageStorage: number[]
}

export interface CreditCardInfo {
  last4: string
  expMonth: number
  expYear: number
  cardHolderName: string
  addressLine1: string
  addressLine2?: string
  addressCity: string
  addressZip: string
  addressState: string
  addressCountry: string
}

export interface ProjectBillingInformation {
  plan: string
  invoices: Invoice[]
}
