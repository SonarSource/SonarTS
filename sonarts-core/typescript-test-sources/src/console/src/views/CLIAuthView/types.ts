export type AuthTrigger = 'auth' | 'init' | 'quickstart'
