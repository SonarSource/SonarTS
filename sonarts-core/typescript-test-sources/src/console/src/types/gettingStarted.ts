import {Example} from './types'

export type Step =
  'STEP0_OVERVIEW' |
  'STEP1_CREATE_POST_MODEL' |
  'STEP2_CLICK_CREATE_FIELD_IMAGEURL' |
  'STEP2_ENTER_FIELD_NAME_IMAGEURL' |
  'STEP2_SELECT_TYPE_IMAGEURL' |
  'STEP2_CLICK_CONFIRM_IMAGEURL' |
  'STEP2_CREATE_FIELD_DESCRIPTION' |
  'STEP2_ENTER_FIELD_NAME_DESCRIPTION' |
  'STEP2_SELECT_TYPE_DESCRIPTION' |
  'STEP2_CLICK_CONFIRM_DESCRIPTION' |

  'STEP3_OPEN_PLAYGROUND' |
  'STEP3_UNCOMMENT_DESCRIPTION' |
  'STEP3_RUN_QUERY1' |
  'STEP3_CREATE_MUTATION_TAB' |
  'STEP3_ENTER_MUTATION1_VALUES' |
  'STEP3_RUN_MUTATION1' |
  'STEP3_ENTER_MUTATION2_VALUE' |
  'STEP3_RUN_MUTATION2' |
  'STEP3_SELECT_QUERY_TAB' |
  'STEP3_RUN_QUERY2' |
  // 'STEP3_OPEN_FINAL_POPUP' |
  // 'STEP3_CLICK_DATA_BROWSER' |
  // 'STEP3_CLICK_ADD_NODE1' |
  // 'STEP3_CLICK_ENTER_IMAGEURL' |
  // 'STEP3_CLICK_ENTER_DESCRIPTION' |
  // 'STEP3_CLICK_SAVE_NODE1' |
  // 'STEP3_CLICK_ADD_NODE2' |
  // 'STEP4_CLICK_PLAYGROUND' |
  // 'STEP4_CLICK_BEGIN_PART1' | // GraphiQL intro
  // 'STEP4_WAITING_PART1' |
  // 'STEP4_CLICK_TEASER_PART2' |
  // 'STEP4_CLICK_BEGIN_PART2' |
  // 'STEP4_WAITING_PART2' |
  // 'STEP4_CLICK_TEASER_STEP5' |

  'STEP5_SELECT_EXAMPLE' |
  'STEP5_WAITING' |
  'STEP5_DONE' |
  'STEP6_CLOSED'

interface StepProgress {
  index: number
  total: number
  done: number
}

interface Props {
  step: Step
  skipped: boolean
  onboardingStatusId: string
  selectedExample?: Example
}

export class GettingStartedState {

  static steps: [Step] = [
    'STEP0_OVERVIEW',
    'STEP1_CREATE_POST_MODEL',
    'STEP2_CLICK_CREATE_FIELD_IMAGEURL',
    'STEP2_ENTER_FIELD_NAME_IMAGEURL',
    'STEP2_SELECT_TYPE_IMAGEURL',
    'STEP2_CLICK_CONFIRM_IMAGEURL',
    'STEP2_CREATE_FIELD_DESCRIPTION',
    'STEP2_ENTER_FIELD_NAME_DESCRIPTION',
    'STEP2_SELECT_TYPE_DESCRIPTION',
    'STEP2_CLICK_CONFIRM_DESCRIPTION',

    'STEP3_OPEN_PLAYGROUND',
    'STEP3_UNCOMMENT_DESCRIPTION',
    'STEP3_RUN_QUERY1',
    'STEP3_CREATE_MUTATION_TAB',
    'STEP3_ENTER_MUTATION1_VALUES',
    'STEP3_RUN_MUTATION1',
    'STEP3_ENTER_MUTATION2_VALUE',
    'STEP3_RUN_MUTATION2',
    'STEP3_SELECT_QUERY_TAB',
    'STEP3_RUN_QUERY2',
    // 'STEP3_OPEN_FINAL_POPUP',
    // 'STEP3_CLICK_DATA_BROWSER',
    // 'STEP3_CLICK_ADD_NODE1',
    // 'STEP3_CLICK_ENTER_IMAGEURL',
    // 'STEP3_CLICK_ENTER_DESCRIPTION',
    // 'STEP3_CLICK_SAVE_NODE1',
    // 'STEP3_CLICK_ADD_NODE2',
    // 'STEP4_CLICK_PLAYGROUND',
    // 'STEP4_CLICK_BEGIN_PART1',
    // 'STEP4_WAITING_PART1',
    // 'STEP4_CLICK_TEASER_PART2',
    // 'STEP4_CLICK_BEGIN_PART2',
    // 'STEP4_WAITING_PART2',
    // 'STEP4_CLICK_TEASER_STEP5',
    // 'STEP5_SELECT_EXAMPLE',
    'STEP5_WAITING',
    'STEP5_DONE',
    'STEP6_CLOSED',
  ]

  step: Step
  skipped: boolean
  onboardingStatusId: string
  selectedExample: Example

  progress: StepProgress

  constructor(props: Props) {
    this.onboardingStatusId = props.onboardingStatusId
    this.selectedExample = props.selectedExample
    this.skipped = props.skipped
    this.update(props.step)
  }

  isActive = (): boolean => {
    return this.step !== 'STEP6_CLOSED' && !this.skipped
  }

  isCurrentStep = (step: Step): boolean => {
    if (this.skipped) {
      return false
    }
    return step === this.step
  }

  update = (step: Step): void => {
    const currentStepIndex = GettingStartedState.steps.indexOf(this.step)
    const stepIndex = GettingStartedState.steps.indexOf(step)
    if (currentStepIndex > stepIndex) {
      return
    }

    this.step = step

    this.progress = {
      'STEP0_OVERVIEW': () => ({ index: 0, total: 0, done: 0 }),

      'STEP1_CREATE_POST_MODEL': () => ({ index: 1, total: 1, done: 0 }),
      'STEP2_CLICK_CREATE_FIELD_IMAGEURL': () => ({ index: 1, total: 2, done: 0 }),
      'STEP2_ENTER_FIELD_NAME_IMAGEURL': () => ({ index: 1, total: 2, done: 0 }),
      'STEP2_SELECT_TYPE_IMAGEURL': () => ({ index: 1, total: 2, done: 0 }),
      'STEP2_CLICK_CONFIRM_IMAGEURL': () => ({ index: 1, total: 2, done: 0 }),
      'STEP2_CREATE_FIELD_DESCRIPTION': () => ({ index: 1, total: 2, done: 1 }),
      'STEP2_ENTER_FIELD_NAME_DESCRIPTION': () => ({ index: 1, total: 2, done: 1 }),
      'STEP2_SELECT_TYPE_DESCRIPTION': () => ({ index: 1, total: 2, done: 1 }),
      'STEP2_CLICK_CONFIRM_DESCRIPTION': () => ({ index: 1, total: 2, done: 1 }),

      'STEP3_OPEN_PLAYGROUND': () => ({index: 2, total: 4, done: 0}),
      'STEP3_UNCOMMENT_DESCRIPTION': () => ({index: 2, total: 4, done: 0}),
      'STEP3_RUN_QUERY1': () => ({index: 2, total: 4, done: 0}),
      'STEP3_CREATE_MUTATION_TAB': () => ({index: 2, total: 4, done: 1}),
      'STEP3_ENTER_MUTATION1_VALUES': () => ({index: 2, total: 4, done: 1}),
      'STEP3_RUN_MUTATION1': () => ({index: 2, total: 4, done: 2}),
      'STEP3_ENTER_MUTATION2_VALUE': () => ({index: 2, total: 4, done: 2}),
      'STEP3_RUN_MUTATION2': () => ({index: 2, total: 4, done: 2}),
      'STEP3_SELECT_QUERY_TAB': () => ({index: 2, total: 4, done: 3}),
      'STEP3_RUN_QUERY2': () => ({index: 2, total: 4, done: 4}),
      // 'STEP3_OPEN_FINAL_POPUP': () => ({index: 3, total: 4, done: 4}),

      // 'STEP3_CLICK_DATA_BROWSER': () => ({ index: 3, total: 2, done: 0 }),
      // 'STEP3_CLICK_ADD_NODE1': () => ({ index: 3, total: 2, done: 0 }),
      // 'STEP3_CLICK_ENTER_IMAGEURL': () => ({ index: 3, total: 2, done: 0 }),
      // 'STEP3_CLICK_ENTER_DESCRIPTION': () => ({ index: 3, total: 2, done: 0 }),
      // 'STEP3_CLICK_SAVE_NODE1': () => ({ index: 3, total: 2, done: 0 }),
      // 'STEP3_CLICK_ADD_NODE2': () => ({ index: 3, total: 2, done: 1 }),
      //
      // 'STEP4_CLICK_PLAYGROUND': () => ({ index: 4, total: 2, done: 0 }),
      // 'STEP4_CLICK_BEGIN_PART1': () => ({ index: 4, total: 2, done: 0 }),
      // 'STEP4_WAITING_PART1': () => ({ index: 4, total:  2, done: 0 }),
      // 'STEP4_CLICK_TEASER_PART2': () => ({ index: 4, total: 2, done: 1 }),
      // 'STEP4_CLICK_BEGIN_PART2': () => ({ index: 4, total: 2, done: 1 }),
      // 'STEP4_WAITING_PART2': () => ({ index: 4, total:  2, done: 1 }),
      // 'STEP4_CLICK_TEASER_STEP5': () => ({ index: 4, total: 2, done: 2 }),

      'STEP5_SELECT_EXAMPLE': () => ({ index: 3, total: 0, done: 0 }),
      'STEP5_WAITING': () => ({ index: 3, total: 0, done: 0 }),
      'STEP5_DONE': () => ({ index: 4, total: 0, done: 0 }),

      'STEP6_CLOSED': () => ({ index: 4, total: 0, done: 0 }),
    }[step]()
  }
}

export interface GettingStartedReducerState {
  poll: boolean,
  gettingStartedState?: GettingStartedState
}
