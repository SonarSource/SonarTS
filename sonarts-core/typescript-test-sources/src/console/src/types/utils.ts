export type NotificationLevel = 'success' | 'error' | 'warning' | 'info'

export interface Notification {
  message: string
  level: NotificationLevel
}
export type ShowNotificationCallback = (notification: Notification) => void

export interface Query {
  query: string
  variables: string
  date: Date
}

export interface NonScalarValue {
  id: string
  [key: string]: TypedValue
}

export type ScalarValue = number | string | Date | boolean
export type AtomicValue = ScalarValue | NonScalarValue
export type TypedValue = ScalarValue | NonScalarValue | Array<AtomicValue>
