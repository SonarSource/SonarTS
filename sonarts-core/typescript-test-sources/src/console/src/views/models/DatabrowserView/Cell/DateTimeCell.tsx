import * as React from 'react'
import Datepicker from '../../../../components/Datepicker/Datepicker'
import {CellProps} from './cells'
import {Moment} from 'moment'

const classes: any = require('../Cell.scss')

export default class DateTimeCell extends React.Component<CellProps<Date>, {}> {

  resolveInput = (m: Moment | string) => {
    if (typeof m === 'string') {
      return new Date(m)
    } else {
      return m.toDate()
    }
  }

  render() {
    return (
      <Datepicker
        className={classes.datepicker}
        defaultValue={this.props.value || new Date()}
        active={true}
        onChange={(m: Moment | string) => this.props.save(this.resolveInput(m))}
        onCancel={() => this.props.cancel()}
        autoClose={this.props.inList}
        defaultOpen={true}
        applyImmediately={false}
        onKeyDown={this.props.onKeyDown}
        onClickOutside={moment => this.props.save(moment.toDate())}
      />)
  }
}
