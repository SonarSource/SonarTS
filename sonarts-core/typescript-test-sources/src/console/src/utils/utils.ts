import * as moment from 'moment'
import {ISO8601} from './constants'
import {Field} from '../types/types'

export function randomString(length: number): string {
  let text = ''
  const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'

  for (let i = 0; i < length; i++) {
    text += possible.charAt(Math.floor(Math.random() * possible.length))
  }

  return text
}

export function isValidUrl(str: string): boolean {
  const pattern = /(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g
  return pattern.test(str)
}

export function isValidMutationCallbackUrl(str: string): boolean {
  const blackList = [
    '127.0.0.1',
    '0.0.0.0',
    'localhost',
  ]
  const withoutProtocol = str.replace(/http(s)?:\/\//, '')
  return !blackList.includes(withoutProtocol) && (isValidUrl(str) || isValidIp(withoutProtocol))
}

export function isValidIp(str: string): boolean {
  const i = str.indexOf('/')
  const withoutSlash = str.slice(0, i)
  const pattern = /^(?!0)(?!.*\.$)((1?\d?\d|25[0-5]|2[0-4]\d)(\.|$)){4}$/
  return pattern.test(withoutSlash)
}

export function isValidDateTime(dateTime: string): boolean {
  // returns whether the string conforms to ISO8601
  // the strict format is '2016-05-19T17:09:24.123Z' but we also accept simpler versions like '2016'
  return moment.utc(dateTime, ISO8601).isValid()
}

export function isValidEnum(value: string): boolean {
  return /^[_a-zA-Z][_a-zA-Z0-9]*$/.test(value)
}

export function debounce(func, wait) {
  let timeout
  return (...args) => {
    const context = this
    const later = () => {
      timeout = null
      func.apply(context, args)
    }
    clearTimeout(timeout)
    timeout = setTimeout(later, wait)
  }
}

interface RetryUntilDoneOptions {
  maxRetries?: number
  timeout?: number
}

export function retryUntilDone(fn: (done: () => void) => void, options: RetryUntilDoneOptions = {}): void {
  const maxRetries = options.maxRetries || 100
  const timeout = options.timeout || 100

  let tries = 0
  let shouldBreak = false

  let interval = setInterval(
    () => {
      tries++

        fn(() => {
        clearInterval(interval)
      })

      if (tries < maxRetries) {
        clearInterval(interval)
      }
    },
    timeout,
  )

}

export function lowercaseFirstLetter(str: string): string {
  return str.charAt(0).toLowerCase() + str.slice(1)
}

export function uppercaseFirstLetter(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

export function removeDuplicatesFromStringArray(arr: string[]): string[] {
  let s = new Set(arr)
  let it = s.values()
  return Array.from(it)
}

// http://stackoverflow.com/questions/2901102/how-to-print-a-number-with-commas-as-thousands-separators-in-javascript
export function numberWithCommas(x) {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
}

export function numberWithCommasRounded(x, digits) {
  return x.toFixed(digits).replace(/\B(?=(\d{3})+(?!\d))/g, ',')
}

export function chunk(str: string, n: number, includeRemainder: boolean): string[] {
  let result: string[] = []
  while (n < str.length) {
    let chunk = str.substring(0, n)
    result.push(chunk)
    str = str.substring(n, str.length)
  }
  if (includeRemainder) {
    result.push(str)
  }
  return result
}

// month is 1 based
export function daysInMonth(month,year) {
  return new Date(year, month, 0).getDate()
}

// input: timestamp with format: "2017-01-31T23:00:00.000Z"
export function mmDDyyyyFromTimestamp(timestamp: string): string {
  const components = timestamp.split('T')
  const dateString = components[0]
  let dateComponents = dateString.split('-')
  const yyyy = dateComponents.shift()
  dateComponents.push(yyyy)
  return dateComponents.join('/')
}

export function todayString(): string {
  const today = new Date()
  const dd = today.getDate()
  const mm = today.getMonth() + 1
  const yyyy = today.getFullYear()
  const todayString = mm + '/' + dd + '/' + yyyy
  return todayString
}

export const idToBeginning = (a: Field, b: Field) => {
  if (a.name === 'id') {
    return -1
  }
  if (b.name === 'id') {
    return 1
  }

  if (a.name === 'description' && b.name === 'imageUrl') {
    return 1
  }

  if (a.name === 'imageUrl' && b.name === 'description') {
    return -1
  }

  return a.name > b.name ? 1 : -1
}
