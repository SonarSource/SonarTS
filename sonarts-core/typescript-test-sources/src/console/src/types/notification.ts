import {NotificationLevel} from '../types/utils'

export interface NotificationState {
  message: string
  level: NotificationLevel
}
