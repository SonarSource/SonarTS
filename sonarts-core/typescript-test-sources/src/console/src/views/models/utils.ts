import {isScalar} from '../../utils/graphql'
import {Field, FieldWidths} from '../../types/types'
import {TypedValue, NonScalarValue, ScalarValue} from '../../types/utils'
import {stringToValue, valueToString, getFieldTypeName} from '../../utils/valueparser'
import calculateSize from 'calculate-size'
import * as Immutable from 'immutable'
import {isNonScalarList} from '../../utils/graphql'

export function emptyDefault(field: Field): TypedValue {

  if (field.isRequired) {
    return null
  }

  const value = function (): any {
    switch (field.typeIdentifier) {
      case 'Int':
        return 0
      case 'Float':
        return 0
      case 'DateTime':
        return new Date().toISOString()
      case 'String':
        return ''
      case 'Boolean':
        return false
      case 'Enum':
        return field.enumValues.length > 0 ? field.enumValues[0] : null
      default:
        return null
    }
  }()

  if (!field.isList) {
    return value
  }

  return []
}

function valueToGQL(value: TypedValue, field: Field): string {
  if (value === null && !field.isRequired) {
    return 'null'
  }

  if (!isScalar(field.typeIdentifier)) {
    if (field.isList) {
      // we can safely assume that it's a list at this point
      const values = value as any[]
      if (values.length === 0) {
        return '[]'
      } else {
        return `[${values.map(v => `"${v.id}"`).join(',')}]`
      }
    }
    return `"${(value as NonScalarValue).id}"`
  }
  if (field.typeIdentifier === 'Enum') {
    if (field.isList) {
      return `[${(value as ScalarValue[]).join(',')}]`
    }
    return value as string
  }

  return JSON.stringify(value)
}

export function toGQL(value: TypedValue, field: Field): string {
  let key = isScalar(field.typeIdentifier) ? field.name : field.isList ? `${field.name}Ids` : `${field.name}Id`

  if (value === null && field.isRequired) {
    return ''
  }

  const output = valueToGQL(value, field)
  return `${key}: ${output}`
}

export function compareFields(a: Field, b: Field): number {
  if (a.name === 'id') {
    return -1
  }
  if (b.name === 'id') {
    return 1
  }
  return a.name.localeCompare(b.name)
}

export function getFirstInputFieldIndex(fields: Field[]): number {
  let inputIndex
  const hasInputField = fields.some((field, index) => {
    if (isNonScalarList(field) || field.name === 'id' || field.isReadonly) {
      return false
    } else {
      inputIndex = index
      return true
    }
  })

  if (hasInputField) {
    return inputIndex
  } else {
    return null
  }
}

export function getDefaultFieldValues(fields: Field[]): { [key: string]: any } {
  return fields.filter((f) => f.name !== 'id')
    .mapToObject(
      (field) => field.name,
      (field) => ({
        value: stringToValue(field.defaultValue, field) || emptyDefault(field),
        field: field,
      }),
    )
}

export function calculateFieldColumnWidths (width: number,
                                            fields: Field[],
                                            nodes: any,
                                           ): FieldWidths {
  const cellFontOptions = {
    font: 'Open Sans',
    fontSize: '12px',
  }

  const headerFontOptions = {
    font: 'Open Sans',
    fontSize: '12px',
  }

  const widths = fields.mapToObject(
    (field) => field.name,
    (field) => {
      switch (field.name) {
        case 'id':
          return 220
        case 'createdAt':
        case 'updatedAt':
          return 200
      }

      let cellWidths = nodes
      .filter(node => !!node)
      .map(node => node.hasOwnProperty('get') ? node.get(field.name) : node[field.name])
      .map(value => valueToString(value, field, false))
      .map(str => calculateSize(str, cellFontOptions).width + 41)
      if (cellWidths.hasOwnProperty('toArray')) {
        cellWidths = cellWidths.toArray()
      }

      const headerWidth = calculateSize(`${field.name} ${getFieldTypeName(field)}`, headerFontOptions).width + 90

      const maxWidth = Math.max(...cellWidths, headerWidth)
      const lowerLimit = field.typeIdentifier === 'DateTime' ? 250 : 150
      const upperLimit = 400

      const idExtra = field.name === 'id' ? 50 : 0
      const relationExtra = field.typeIdentifier === 'Relation' ? 50 : 0

      return (
        maxWidth > upperLimit ?
        upperLimit :
        (maxWidth < lowerLimit ? lowerLimit : maxWidth)
      ) + idExtra + relationExtra
    },
  )

  const totalWidth = fields.reduce((sum, {name}) => sum + widths[name], 0)
  const fieldWidth = width - 34 - 250
  if (totalWidth < fieldWidth) {
    fields.forEach(({name}) => {
      if (!['id', 'createdAt', 'updatedAt'].includes(name)) {
        widths[name] = (widths[name] / totalWidth) * fieldWidth
      }
    })
  }
  return widths
}
