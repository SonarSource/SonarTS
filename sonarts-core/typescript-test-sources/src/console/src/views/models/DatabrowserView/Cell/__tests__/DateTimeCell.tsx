import * as React from 'react' // tslint:disable-line
import DateTimeCell from '../DateTimeCell'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import {FieldType} from '../../../../../types/types'
import {TypedValue} from '../../../../../types/utils'

test('DateTimeCell renders', () => {

  const save = jest.fn((value: TypedValue) => { /* */ })
  const cancel = jest.fn()
  const onKeyDown = jest.fn()
  const field = {
    id: 'cip3p48sj001e1jsmghwkdt2k',
    name: 'description',
    description: '',
    isReadonly: true,
    isList: false,
    isSystem: true,
    typeIdentifier: 'DateTime' as FieldType,
    relatedModel: null,
    isUnique: true,
    isRequired: false,
    enumValues: [],
    model: null,
  }

  const component = shallow(
    <DateTimeCell
      value={new Date('Sat Oct 15 2016 20:09:46 GMT+0200 (CEST)')}
      save={save}
      cancel={cancel}
      onKeyDown={onKeyDown}
      field={field}
    />,
  )

  expect(shallowToJson(component)).toMatchSnapshot()

})
