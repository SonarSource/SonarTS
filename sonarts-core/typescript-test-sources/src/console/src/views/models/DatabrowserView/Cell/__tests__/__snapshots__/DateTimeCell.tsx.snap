exports[`test DateTimeCell renders 1`] = `
<DatePicker
  active={true}
  applyImmediately={false}
  className="datepicker"
  defaultOpen={true}
  defaultValue={2016-10-15T18:09:46.000Z}
  onCancel={[Function]}
  onChange={[Function]}
  onClickOutside={[Function]}
  onKeyDown={[Function]} />
`;
